---
title: "Smed CisReg project: Planarian HNF4 knock-down and Differential Gene Expression; library preparation"
author: "Alberto Perez-Posada @apposada"
date: "8/11/2023"
output:
  html_document: default
  md_document:
    variant: markdown_github
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}

dir <- '/mnt/sda/alberto/projects/smed_cisreg/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)

```

## About

This markdown attempts to recapitulate how the matrices were loaded and merged in R.

## Load libraries

```{r libraries, message=FALSE, warning=FALSE}
library(vroom)
library(plyr)
library(dplyr)
library(ggplot2)
library(Seurat)
library(Matrix)
```

## Load functions

```{r load_functions}
source("code/r_code/functions/sourcefolder.R")

sourceFolder(
  "code/r_code/functions/", #change for definitive path once its final
  recursive = TRUE
  )
```

## Load data

We will load each of these matrices as separate Seurat objects. (This will take a lot of time.)

```{r load_matrices}
L113.data <- 
  read.table(
    file = "outputs/scrna/L11.3/L11.3_allCells.txt.gz",
    header = TRUE, row.names = 1, colClasses = c("character", rep("numeric"))
    )
L114.data <- 
  read.table(
    file = "outputs/scrna/L11.4/L11.4_allCells.txt.gz",
    header = TRUE, row.names = 1, colClasses =c("character", rep("numeric"))
    )
```

Creation of a Seurat object for each:

```{r CreateSeuratObject}
#Create a seurat object for each file.
L113 <- CreateSeuratObject(counts = L113.data, project = "11_3", min.cells = 1, min.features = 100)
L114 <- CreateSeuratObject(counts = L114.data, project = "11_4", min.cells = 1, min.features = 100)
```

We will normalise them and tag the cells with their respective matrix of origin and, after that, we will merge all the matrices into one single matrix.

```{r normalise seuratobjects}
# PFlog1pPF normalisation
L113@assays$RNA@data <- do_pf(log1p(do_pf(L113@assays$RNA@data)))
L114@assays$RNA@data <- do_pf(log1p(do_pf(L114@assays$RNA@data)))
```

Here we must add the calculation of variable genes on each separate library. We have a large number of genic features in this genome annotation (around 50K, including hi-confidence and "other"); therefore we amp up the number of variable genes.

```{r variablegenes}
#Find variable features
L113 <- FindVariableFeatures(L113, selection.method = "vst", nfeatures = 20000)
L114 <- FindVariableFeatures(L114, selection.method = "vst", nfeatures = 20000)
```

```{r assign_library_identity}
#Assign an identity to each dataset
L113[["library"]] <- "11_3"
L114[["library"]] <- "11_4"

L113@meta.data$library <- "11_3"
L114@meta.data$library <- "11_4"
```

The merge:

```{r merge_seuratobjects}
add.cell.ids <- c("11_3", "11_4")

smed_hnf_scrna <- merge(x = L113, y = list(L114), add.cell.ids = add.cell.ids, merge.data = FALSE)

Idents(smed_hnf_scrna) <- "library"  # use identity based on sample identity
```

Here is a look at how the number of genes per cell varies across the different libraries.

```{r VlnPlot_nFeat}
ComparisonLibrary_nfeat <- VlnPlot(smed_hnf_scrna, "nFeature_RNA", group.by = "library")+ggtitle("_Smed_HNF_nFeature_By_Library_highgene_cutoff.pdf")
print(ComparisonLibrary_nfeat)
```

And here is a look at how the number of counts detected per cell varies across the different libraries.

```{r VlnPlot_nCount}
ComparisonLibrary_ncount <- VlnPlot(smed_hnf_scrna, "nCount_RNA", group.by = "library")+ggtitle("_Smed_HNF_nCount_By_Library_highgene_cutoff.pdf")
print(ComparisonLibrary_ncount)
```

We will normalise the concatenated data now, and subset to remove outlier cells.

```{r normalisation_and_subset}
# normalized merged data
smed_hnf_scrna@assays$RNA@data <-
  do_pf(log1p(do_pf(smed_hnf_scrna@assays$RNA@data)))

# subset
smed_hnf_scrna <- 
  subset(
    smed_hnf_scrna, 
    subset = nFeature_RNA < 750 & nCount_RNA < 1500
    )
```

A quick plot for visualisation:

```{r}
BasicFeatures_presubset <- 
  VlnPlot(smed_hnf_scrna, features = c("nFeature_RNA", "nCount_RNA"))

print(BasicFeatures_presubset)
```

A quick overview to this object:

```{r overview}
smed_hnf_scrna
```

We will finally save this file as an object we can load later on:

```{r save scrna}
#Save
saveRDS(
  smed_hnf_scrna,
  file = "outputs/rda/seurat_scrna/hnf4_rnai/smed_hnf_scrna_concatenated_libraries.RDS"
)
```
