---
title: "Smed CisReg project: Downstream analysis of HNF4 K.D. Differentially Expressed genes. Targets in Network and Motif Analysis"
author: "Alberto Perez-Posada @apposada"
date: "08/05/2024"
output: html_document
---

```{r setup, include=FALSE}

dir <- '/mnt/sda/alberto/projects/smed_cisreg/'
fcha <- function(){ gsub("-","",Sys.Date()) }

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = dir)
options(scipen=999)
Sys.setenv(VROOM_CONNECTION_SIZE=5000072)
pdf.options(family = "Helvetica")
```

## About

This markdown helps in Differential Gene Expression Analysis using DESeq2. This is a much simpler, abridged version of Maria Rosello's original markdown file which was the basis for this entire markdown. I reccommend checking her markdown as well as it delves much deeper into data structure and normalisation, as well as provides a more in-detail explanation and an additional method for DGE analysis (the package `limma`).

## Load libraries

We will need these libraries for the markdown to run.

```{r libraries, message=FALSE, warning=FALSE}
require(DESeq2)
require(dplyr)
require(ggplot2)
library(ggrepel)
library(ggvenn)
require(ComplexHeatmap)
library(topGO)
library(readr)
library(openxlsx)
library(colorspace)
library(gridExtra)
library(ggrastr)
```

## Load functions

We will use here a simple wrapper of DESeq and a couple of plotting functions. The way this works is:

 - It takes the a count matrix and a samples information matrix to perform DGE analysis.
 - If no DGEs are found, it ends with a message indicating so.
 - If it only finds one DGE, it reports its ID.
 - If it finds more than one DGE, it proceeds to plot a volcano plot and a heatmap of normalised gene expression.
 - In the two latter cases, a list of things is returned: the DESeq2 object which contains the whole DGE analysis, the results table from running DESeq2, and a list of differentially expressed genes.
 - If plots were made, they will also be in the output of the function.

In the future, this function will allow to select what type of p to filter by: p-value or adjusted p-value, as well as a LFC threshold.

```{r load_functions}
source("~/projects/smed_cisreg/code/r_code/functions/sourcefolder.R")
sourceFolder("code/r_code/functions/",recursive = FALSE)
```

## Load data

```{r load_data}
#smed_hnf_scrna <- readRDS("outputs/rda/seurat_scrna/hnf4_rnai/smed_hnf_scrna_SeuratObject.RDS")

smed_ctypes <- 
  read.delim2(
    file = "~/projects/smed_cisreg/outputs/smed_cisreg_ctypes.tsv",
    sep = "\t",
    header = TRUE
    )
smed_ctypes <- unique(smed_ctypes)

rosetta <- read.delim2(file = "~/../Standard_References/Smed_Rink/20231127_Rosetta.tsv", header = TRUE)

load("~/projects/smed_cisreg/outputs/rda/smed_hnf_analysis.rda")
```

We load the whole networks of phagocyte and parenchyma too:

```{r}
# phagocytes
pha_nw <- data.table::fread("outputs/ananse/outs/network/phagocytes.network", header = TRUE)
pha_nw <- as.data.frame(pha_nw)
pha_nw$tf <- sub("—.*","",pha_nw$tf_target)
pha_nw$tg <- sub(".*—","",pha_nw$tf_target)
pha_nw <- pha_nw[,c(7,8,2,4,5,6)]

# parenchyma
par_nw <- data.table::fread("outputs/ananse/outs/network/parenchyma.network", header = TRUE)
par_nw <- as.data.frame(par_nw)
par_nw$tf <- sub("—.*","",par_nw$tf_target)
par_nw$tg <- sub(".*—","",par_nw$tf_target)
par_nw <- par_nw[,c(7,8,2,4,5,6)]
```

## ANANSE scores

```{r}
degs_pha <- 
  hnf_DGE_all_broad$phagocytes$diffgenes[
    !(hnf_DGE_all_broad$phagocytes$diffgenes %in%
        hnf_DGE_all_broad$parenchyma$diffgenes)
  ]
degs_par <-
  hnf_DGE_all_broad$parenchyma$diffgenes[
    !(hnf_DGE_all_broad$parenchyma$diffgenes %in%
        hnf_DGE_all_broad$phagocytes$diffgenes)
  ]
degs_both <- 
  hnf_DGE_all_broad$phagocytes$diffgenes[
    hnf_DGE_all_broad$phagocytes$diffgenes %in%
      hnf_DGE_all_broad$parenchyma$diffgenes
  ]

deg_cols <- setNames(c("#c4c4c4","#D675AD","#d39898","#53cc86"),c("none","par","both","pha"))
```

We keep only the targets of HNF4:

```{r}
# DF PHA
df_pha <- pha_nw[pha_nw$tf == hnf4, -1]
# DF PAR
df_par <- par_nw[par_nw$tf == hnf4, -1]
```

We parse this table of target genes and add them the info of whether they are downregulated in one, the other, or both cell types. Plus we rank the interaction score (which is an additive model accounting binding and co-expression) and the weighted binding (which does not account co-expression).

```{r}
# DEGs pha
df_pha$deg <- "none"
df_pha$deg[df_pha$tg %in% degs_pha] <- "pha"
# DEGs par
df_pha$deg[df_pha$tg %in% degs_par] <- "par"
# DEGs both
df_pha$deg[df_pha$tg %in% degs_both] <- "both"
df_pha$deg <- factor(df_pha$deg, levels = c("none","par","both","pha"))

df_pha$rank_prob <- relativise(rank(df_pha$prob, ties.method = "min"))
df_pha$rank_wb <- relativise(rank(df_pha$weighted_binding, ties.method = "min"))

head(df_pha)
```

We do the same for the parenchyma network:

```{r}
df_par$deg <- "none"
# DEGs pha
df_par$deg[df_par$tg %in% degs_pha] <- "pha"
# DEGs par
df_par$deg[df_par$tg %in% degs_par] <- "par"
# DEGs both
df_par$deg[df_par$tg %in% degs_both] <- "both"
df_par$deg <- factor(df_par$deg, levels = c("none","par","both","pha"))

df_par$rank_prob <- relativise(rank(df_par$prob, ties.method = "min"))
df_par$rank_wb <- relativise(rank(df_par$weighted_binding, ties.method = "min"))
head(df_par)
```

Here is the boxplot:

```{r, fig.width = 4.5, fig.height = 4.5}
set.seed(123)
non_deg_subsample_pha <- sample(df_pha$tg[df_pha$deg == "none"], 500)
non_deg_subsample_par <- sample(df_par$tg[df_pha$deg == "none"], 500)

f_pha <- which(df_pha$tg %in% c(non_deg_subsample_pha,df_pha$tg[df_pha$deg != "none"]))
f_par <- which(df_par$tg %in% c(non_deg_subsample_par,df_par$tg[df_par$deg != "none"]))

deg_outl_col <- deg_cols
deg_outl_col <- darken(deg_outl_col,.7)
deg_outl_col[1] <- alpha(deg_outl_col[1],.2)
deg_outl_col[c(2,3,4)] <- alpha(deg_outl_col[c(2,3,4)],.25)

set.seed(1234)
par(mfrow = c(1,2))
boxplot(
  df_pha$prob~df_pha$deg,
  col = deg_cols, border = darken(deg_cols,.7),
  main = "Interaction score\nof HNF4 predicted targets\n(phagocytes)",
  xlab = "DEGs in Hnf4 K.D.",
  ylab = "Interaction score",
  las = 2,
  cex.main = .6,
  frame.plot = FALSE,
  outline = FALSE
)
stripchart(
  df_pha$prob[f_pha]~df_pha$deg[f_pha],
  col = deg_outl_col,
  method = "jitter",
  jitter=0.15,
  vertical = TRUE,
  pch = 20,
  cex=0.7,
  add = TRUE
)
boxplot(
  df_par$prob~df_par$deg,
  col = deg_cols, border = darken(deg_cols,.7),
  main = "Interaction score\nof HNF4 predicted targets\n(parenchyma)",
  xlab = "DEGs in Hnf4 K.D.",
  ylab = "Interaction score",
  las = 2,
  cex.main = .6,
  frame.plot = FALSE,
  outline = FALSE
)
stripchart(
  df_par$prob[f_par]~df_par$deg[f_par],
  col = deg_outl_col,
  method = "jitter",
  jitter=0.15,
  vertical = TRUE,
  pch = 20,
  cex=0.7,
  add = TRUE
)
par(mfrow = c(1,1))
```


We will merge these two data frames to plot a scatter of these target genes based on their interaction with HNF4 in phagocytes or parenchyma.

```{r}
# Merge for scatter
df_pha_par <-
  merge(
    df_pha,
    df_par,
    by.x = 1,
    by.y = 1,
    all.x = TRUE,
    all.y = TRUE,
  )

df_pha_par[is.na(df_pha_par)] <- 0

df_pha_par$col <-
  rgb(.1,.1,.1,.005)

pha_bool <- df_pha_par$tg %in% hnf_DGE_all_broad$phagocytes$diffgenes
par_bool <- df_pha_par$tg %in% hnf_DGE_all_broad$parenchyma$diffgenes
```

Here the scatter:

```{r, fig.width = 4.5, fig.height = 4.5}
plot(
  df_pha_par$rank_prob.x,
  df_pha_par$rank_prob.y,
  pch = 20,
  col = rgb(.2,.2,.2,.05),
  xlab = "Interaction Score Phagocytes",
  ylab = "Interaction Score parenchyma",
  main = "ANANSE interaction of HNF4 and targets"
)

points(
  df_pha_par$rank_prob.x[pha_bool],
  df_pha_par$rank_prob.y[pha_bool],
  pch = 20,
  cex = 1.1,
  col = "#53cc86"
)

points(
  df_pha_par$rank_prob.x[par_bool],
  df_pha_par$rank_prob.y[par_bool],
  pch = 20,
  cex = 1.1,
  col = "#D675AD"
)

points(
  df_pha_par$rank_prob.x[pha_bool & par_bool],
  df_pha_par$rank_prob.y[pha_bool & par_bool],
  pch = 20,
  cex = 1.1,
  col = "#d39898"#"#C27D65"
)

legend(
  "bottomright",
  legend = c("DEG phagocytes","DEG parenchyma", "DEG both", "none"),
  col = c("#53cc86","#D675AD","#d39898",rgb(.2,.2,.2,.5)),
  pch = 20,
  cex = 1.1
)
par(mfrow = c(1,1))
```


## Motif analysis

We will write down the promoters of these groups of DEGs to make a motif enrichment analysis.

```{r}
hnf_degs_df <- data.frame(
  id = sort(unique(unlist(hnf_DEG_phpr_pgrn))),
  in_pha = ifelse(sort(unique(unlist(hnf_DEG_phpr_pgrn))) %in% hnf_DEG_phpr_pgrn$phagocytes, 1, 0),
  in_par = ifelse(sort(unique(unlist(hnf_DEG_phpr_pgrn))) %in% hnf_DEG_phpr_pgrn$parenchyma, 1, 0)
)
hnf_degs_df$in_both <- ifelse(hnf_degs_df$in_pha+hnf_degs_df$in_par == 2, 1, 0)

smed_promoters <- read.delim2(file = "/mnt/sda/alberto/projects/smed_cisreg/outputs/associate_peaks_genes/promoters.bed", header = FALSE)

write.table(
  smed_promoters[smed_promoters$V4 %in% hnf_degs_df$id[hnf_degs_df$in_pha == 1],],
  "outputs/dge/hnf4/homer/promoters/smed_promoters_pha_deg.bed",
  sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE
)

write.table(
  smed_promoters[smed_promoters$V4 %in% hnf_degs_df$id[hnf_degs_df$in_par == 1],],
  "outputs/dge/hnf4/homer/promoters/smed_promoters_par_dge.bed",
  sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE
)

write.table(
  smed_promoters[smed_promoters$V4 %in% hnf_degs_df$id[hnf_degs_df$in_pha == 1],],
  "outputs/dge/hnf4/homer/promoters/smed_promoters_both_deg.bed",
  sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE
)

write.table(
  smed_promoters[smed_promoters$V4 %in% hnf_degs_df$id,],
  "outputs/dge/hnf4/homer/smed_promoters_deg_all.bed",
  sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE
)
```


We performed motif enrichment analysis of the promoters of these DEGs. Here we will plot these results. We load the table and parse it to keep those motifs with q-value < 0.1

```{r}
#consider using -h option in homer as well as -mis 3 instead of -mis 2
# Load table
motifs_degs_0 <- 
  read_tsv("outputs/dge/hnf4/homer/results_bg_allpromotersingenome//motifs_all.tsv")

# tidyr wrapper to present a tidy table
motifs_degs <- parse_homer_output_table(
  tsv_motifs = motifs_degs_0,
  categ_regex1 = "homer_output_promoters_module_",
  categ_regex2 = "",
  qval_thresh = 0.1,
  max_logqval = 10
)
colnames(motifs_degs)[1] <- "module"
```

We parse this table a bit to tidy it up, and give the motifs a proper ordering etc.

```{r}
# Transform module names column into ordered factor much like other plots from the WGCNA analysis
motifs_degs$module <- 
  factor(motifs_degs$module, levels = rev(c("pha","both","par")))

# Arrange by module name and decreasing values of % peaks with motif found, to facilitate readability
motifs_degs <- 
  motifs_degs[
    with(motifs_degs,order(module,-as.numeric(pct_target_seqs_with_motif))),
  ]#https://stackoverflow.com/questions/16205232/order-data-frame-by-columns-in-increasing-and-decreasing-order

# Transform motifs names column into ordered factor so that ggplot respects the ordering we did in the step immediately above
motifs_degs$motif <- 
  factor(motifs_degs$motif, levels = rev(unique(motifs_degs$motif)))
```

Here the plot for all significantly enriched motifs

```{r, fig.height = 3, fig.width = 4}
p_mot_degs <- 
  ggplot(
    data = motifs_degs, 
    aes(x = motif, y = module,
        color = `logqval`, size = as.factor(size_pct))) +
  geom_point() +
  scale_color_gradient(
    low = alpha("#401af0",0.7), high = alpha("#be0143",0.7)
    ) +
  theme_bw() +
  xlab("") + ylab("") +
  ggtitle("Motif enrichment in DEGs") +
  theme(
    text = element_text(size=10), legend.text = element_text(size=10), 
    axis.text.x = element_text(angle = 90, vjust = 0.5)
    ) +
  scale_size_manual(values = c(1.5, 2.5, 3.5, 4.5,5.5)*1.5, labels = c("1-10", "10-20", "20-50", "50+")) +
  guides(size=guide_legend(title="% peaks with motif"))

print(p_mot_degs)

pdf("graphics/hnf_motif.pdf", height = 3, width = 4)
print(p_mot_degs)
dev.off()
```

These motifs correspond to nuclear receptor (such as HNF4), Homeobox (such as NKX), Stat, and FoxD (such as FoxF).

These correspond to TFs that are mostly specific to 

## Single cell data umaps 

```{r}
scdata <- readRDS("outputs/rda/seurat_scrna/smed_cisreg_scRNA_consensus_identities.RDS")

hnf4 <- "h1SMcG0019688"
nkx2 <- "h1SMcG0009317"
fox2 <- "h1SMcG0016896"

hnf_umap <-
  FeaturePlot(
    scdata,
    features = hnf4,
    cols = c("#DAE7F2","#531ccb"),
    order = TRUE,
    pt.size = 1
  )+NoAxes()+NoLegend()+ggtitle("")

nkx2_umap <-
  FeaturePlot(
    scdata,
    features = nkx2,
    cols = c("#DAE7F2","#228B22"),
    order = TRUE,
    pt.size = 1
  )+NoAxes()+NoLegend()+ggtitle("")

foxf_umap <-
  FeaturePlot(
    scdata,
    features = fox2,
    cols = c("#DAE7F2","#d82880"),
    order = TRUE,
    pt.size = 1
  )+NoAxes()+NoLegend()+ggtitle("")
p_clu_i <- CombinePlots(plots = list(nkx2_umap, hnf_umap, foxf_umap),ncol = 3)
gridExtra::grid.arrange(p_clu_i)
```

```{r}
load("outputs/rda/wgcna_g_merge_aggr_2plus.rda")
load("outputs/rda/tf_analysis.rda")
pal <- c("white", sequential_hcl(5, "YlOrRd", rev = T))
give_col <- circlize::colorRamp2(seq(.4,1, len = length(pal)),pal)
V(g_merge_aggr_2plus)$nkx_col <- give_col(relativise(tf_eigen[nkx2,])[match(V(g_merge_aggr_2plus)$name,names(tf_eigen[nkx2,]))])
V(g_merge_aggr_2plus)$hnf_col <- give_col(relativise(tf_eigen[hnf4,])[match(V(g_merge_aggr_2plus)$name,names(tf_eigen[hnf4,]))])
V(g_merge_aggr_2plus)$fox2_col <- give_col(relativise(tf_eigen[fox2,])[match(V(g_merge_aggr_2plus)$name,names(tf_eigen[fox2,]))])
```

```{r}
par(mfrow = c(1,3))
plot(
  main = "NKX connectivity",
  g_merge_aggr_2plus, 
  edge.color = rgb(0,0,0,.1),
  vertex.label = NA,
  vertex.color = V(g_merge_aggr_2plus)$nkx_col,
  vertex.frame.color = darken(V(g_merge_aggr_2plus)$nkx_col, 0.5),
  mark.groups = split(names(membership(clp)),membership(clp)),
  mark.col= alpha(c("#ea9999","#84aed3","#a0db8e","#fbe5a0","#A88E49","#ffeadd","#d9d2e9"),.1),
  mark.border=NA,
  layout = l,
  edge.width = E(g_merge_aggr_2plus)$w2,
  vertex.label.dist=0.5,
  vertex.label.cex = 0.8,
  vertex.size = 6
)

plot(
  main = "HNF connectivity",
  g_merge_aggr_2plus, 
  edge.color = rgb(0,0,0,.1),
  vertex.label = NA,
  vertex.color = V(g_merge_aggr_2plus)$hnf_col,
  vertex.frame.color = darken(V(g_merge_aggr_2plus)$hnf_col, 0.5),
  mark.groups = split(names(membership(clp)),membership(clp)),
  mark.col= alpha(c("#ea9999","#84aed3","#a0db8e","#fbe5a0","#A88E49","#ffeadd","#d9d2e9"),.1),
  mark.border=NA,
  layout = l,
  edge.width = E(g_merge_aggr_2plus)$w2,
  vertex.label.dist=0.5,
  vertex.label.cex = 0.8,
  vertex.size = 6
)

plot(
  main = "FOXF connectivity",
  g_merge_aggr_2plus, 
  edge.color = rgb(0,0,0,.1),
  vertex.label = NA,
  vertex.color = V(g_merge_aggr_2plus)$fox2_col,
  vertex.frame.color = darken(V(g_merge_aggr_2plus)$fox2_col, 0.5),
  mark.groups = split(names(membership(clp)),membership(clp)),
  mark.col= alpha(c("#ea9999","#84aed3","#a0db8e","#fbe5a0","#A88E49","#ffeadd","#d9d2e9"),.1),
  mark.border=NA,
  layout = l,
  edge.width = E(g_merge_aggr_2plus)$w2,
  vertex.label.dist=0.5,
  vertex.label.cex = 0.8,
  vertex.size = 6
)
par(mfrow = c(1,1))
```


## Saving the plots

```{r}
pdf("graphics/hnf_degs_targetgenes.pdf", width = 4.5, height = 4.5)
set.seed(1234)
par(mfrow = c(1,2))
boxplot(
  df_pha$prob~df_pha$deg,
  col = deg_cols, border = darken(deg_cols,.7),
  main = "Interaction score\nof HNF4 predicted targets\n(phagocytes)",
  xlab = "DEGs in Hnf4 K.D.",
  ylab = "Interaction score",
  las = 2,
  cex.main = .6,
  frame.plot = FALSE,
  outline = FALSE
)
stripchart(
  df_pha$prob[f_pha]~df_pha$deg[f_pha],
  col = deg_outl_col,
  method = "jitter",
  jitter=0.15,
  vertical = TRUE,
  pch = 20,
  cex=0.7,
  add = TRUE
)
boxplot(
  df_par$prob~df_par$deg,
  col = deg_cols, border = darken(deg_cols,.7),
  main = "Interaction score\nof HNF4 predicted targets\n(parenchyma)",
  xlab = "DEGs in Hnf4 K.D.",
  ylab = "Interaction score",
  las = 2,
  cex.main = .6,
  frame.plot = FALSE,
  outline = FALSE
)
stripchart(
  df_par$prob[f_par]~df_par$deg[f_par],
  col = deg_outl_col,
  method = "jitter",
  jitter=0.15,
  vertical = TRUE,
  pch = 20,
  cex=0.7,
  add = TRUE
)
par(mfrow = c(1,1))
dev.off()
```



```{r}
pdf("graphics/scatter_all_hnf4_tgs.pdf", width = 4.5, height = 4.5)
plot(
  df_pha_par$rank_prob.x,
  df_pha_par$rank_prob.y,
  pch = 20,
  col = rgb(.2,.2,.2,.05),
  xlab = "Interaction Score Phagocytes",
  ylab = "Interaction Score parenchyma",
  main = "ANANSE interaction of HNF4 and targets"
)

points(
  df_pha_par$rank_prob.x[pha_bool],
  df_pha_par$rank_prob.y[pha_bool],
  pch = 20,
  cex = 1.1,
  col = "#53cc86"
)

points(
  df_pha_par$rank_prob.x[par_bool],
  df_pha_par$rank_prob.y[par_bool],
  pch = 20,
  cex = 1.1,
  col = "#D675AD"
)

points(
  df_pha_par$rank_prob.x[pha_bool & par_bool],
  df_pha_par$rank_prob.y[pha_bool & par_bool],
  pch = 20,
  cex = 1.1,
  col = "#d39898"#"#C27D65"
)

legend(
  "bottomright",
  legend = c("DEG phagocytes","DEG parenchyma", "DEG both", "none"),
  col = c("#53cc86","#D675AD","#d39898",rgb(.2,.2,.2,.5)),
  pch = 20,
  cex = 1.1
)
par(mfrow = c(1,1))
dev.off()
```

```{r}

png(
  file = "graphics/hnf_nkx2_foxf_umaps.png",
  width= 1152,
  height = 384
)
gridExtra::grid.arrange(p_clu_i)
dev.off()

pdf("graphics/hnf4_connectivity_tfs_in_network.pdf", width = 9, height = 3)
par(mfrow = c(1,3))
plot(
  main = "NKX connectivity",
  g_merge_aggr_2plus, 
  edge.color = rgb(0,0,0,.1),
  vertex.label = NA,
  vertex.color = V(g_merge_aggr_2plus)$nkx_col,
  vertex.frame.color = darken(V(g_merge_aggr_2plus)$nkx_col, 0.5),
  mark.groups = split(names(membership(clp)),membership(clp)),
  mark.col= alpha(c("#ea9999","#84aed3","#a0db8e","#fbe5a0","#A88E49","#ffeadd","#d9d2e9"),.1),
  mark.border=NA,
  layout = l,
  edge.width = E(g_merge_aggr_2plus)$w2,
  vertex.label.dist=0.5,
  vertex.label.cex = 0.8,
  vertex.size = 8.5
)

plot(
  main = "HNF connectivity",
  g_merge_aggr_2plus, 
  edge.color = rgb(0,0,0,.1),
  vertex.label = NA,
  vertex.color = V(g_merge_aggr_2plus)$hnf_col,
  vertex.frame.color = darken(V(g_merge_aggr_2plus)$hnf_col, 0.5),
  mark.groups = split(names(membership(clp)),membership(clp)),
  mark.col= alpha(c("#ea9999","#84aed3","#a0db8e","#fbe5a0","#A88E49","#ffeadd","#d9d2e9"),.1),
  mark.border=NA,
  layout = l,
  edge.width = E(g_merge_aggr_2plus)$w2,
  vertex.label.dist=0.5,
  vertex.label.cex = 0.8,
  vertex.size = 8.5
)

plot(
  main = "FOXF connectivity",
  g_merge_aggr_2plus,
  vertex.label = NA,
  edge.color = rgb(0,0,0,.1),
  vertex.color = V(g_merge_aggr_2plus)$fox2_col,
  vertex.frame.color = darken(V(g_merge_aggr_2plus)$fox2_col, 0.5),
  mark.groups = split(names(membership(clp)),membership(clp)),
  mark.col= alpha(c("#ea9999","#84aed3","#a0db8e","#fbe5a0","#A88E49","#ffeadd","#d9d2e9"),.1),
  mark.border=NA,
  layout = l,
  edge.width = E(g_merge_aggr_2plus)$w2,
  vertex.label.dist=0.5,
  vertex.label.cex = 0.8,
  vertex.size = 8.5
)
par(mfrow = c(1,1))
dev.off()


lgd <- Legend(col_fun = give_col, title = "Relative\nconnectivity", border = "#4c4c4c")
pdf("graphics/hnf_g_aggr_legend.pdf",width = 1.5, height = 1.5)
draw(lgd)
dev.off()
```
